File: haskell/stm/STMBank.hs
-----------------------------------------------------------------------------
-- STMBank.hs -- STM (Software Transactional Memory) Bank Transfer
--
-- Paradigm: Composable concurrent transactions via Software Transactional
--           Memory (STM)
--
-- GHC's STM provides composable atomic transactions using TVar (transactional
-- variables). Unlike lock-based approaches, STM operations are:
--   1. Composable  -- combine multiple TVar reads/writes atomically
--   2. Safe        -- no deadlocks or race conditions
--   3. Retry-based -- transactions retry automatically on conflict
--
-- This demo creates 5 bank accounts and spawns 7 concurrent threads that
-- perform random transfers. STM guarantees that each transfer is atomic
-- and that the total balance is always conserved.
--
-- Build:
--   ghc -O2 -threaded -package stm -o STMBank STMBank.hs
--
-- Run:
--   ./STMBank
-----------------------------------------------------------------------------

module Main where

import Control.Concurrent (forkIO)
import Control.Concurrent.MVar (newEmptyMVar, putMVar, takeMVar)
import Control.Concurrent.STM (atomically, STM)
import Control.Concurrent.STM.TVar (TVar, modifyTVar', newTVarIO, readTVar, writeTVar)
import Control.Monad (replicateM)
import System.Random (randomRIO)
import Text.Printf (printf)

-----------------------------------------------------------------------------
-- Core types and operations
-----------------------------------------------------------------------------

-- | A bank account is a 'TVar Int' holding the current balance.
-- TVar ensures that all reads and writes participate in an STM transaction.
type Account = TVar Int

-- | Create a new account with the given initial balance.
newAccount :: Int -> IO Account
newAccount = newTVarIO

-- | Read the current balance in a transaction.
readBalance :: Account -> STM Int
readBalance = readTVar

-- | Print the balance of a single account (IO action, not transactional).
printBalance :: Account -> IO Int
printBalance acc = atomically $ readTVar acc

-----------------------------------------------------------------------------
-- Transfer operations -- demonstrating STM composability
-----------------------------------------------------------------------------

-- | Atomic transfer between two accounts.
-- Both operations (debit and credit) happen in a single STM transaction.
-- No other thread can observe an intermediate state where money has been
-- debited but not yet credited.
--
-- If you tried to do this with separate locks, you would need to acquire
-- locks on both accounts, which risks deadlock. STM eliminates this problem.
transfer :: Account -> Account -> Int -> IO ()
transfer from to amount = atomically $ do
    fromBal <- readTVar from
    if fromBal < amount
        then error "Insufficient funds"
        else do
            writeTVar from (fromBal - amount)
            modifyTVar' to (+ amount)

-- | Safe version of transfer that returns a success flag instead of crashing.
-- This is still a single, atomic, composable transaction.
safeTransfer :: Account -> Account -> Int -> IO Bool
safeTransfer from to amount = atomically $ do
    fromBal <- readTVar from
    if fromBal < amount
        then return False
        else do
            writeTVar from (fromBal - amount)
            modifyTVar' to (+ amount)
            return True

-- | Atomically compute the total balance across all accounts.
-- This is a composite STM operation -- it reads multiple TVars in a single
-- transaction, guaranteeing a consistent snapshot of the system.
totalBalance :: [Account] -> IO Int
totalBalance accounts = atomically $ do
    balances <- mapM readTVar accounts
    return $ sum balances

-----------------------------------------------------------------------------
-- Transfer thread
-----------------------------------------------------------------------------

-- | Run a single thread that performs 'n' random transfers.
transferThread :: [Account] -> Int -> IO ()
transferThread accounts = go
  where
    numAccounts = length accounts

    go 0 = pure ()
    go n = do
        fromIdx <- randomRIO (0, numAccounts - 1)
        toIdx   <- randomRIO (0, numAccounts - 1)
        amt     <- randomRIO (10, 200)
        if fromIdx /= toIdx
            then do
                let fromAcc = accounts !! fromIdx
                    toAcc   = accounts !! toIdx
                _ <- safeTransfer fromAcc toAcc amt
                go (n - 1)
            else
                go n  -- retry with different accounts

-----------------------------------------------------------------------------
-- Helpers
-----------------------------------------------------------------------------

printBalances :: [Account] -> IO ()
printBalances accounts = do
    balances <- mapM printBalance accounts
    let labeled = zip [1 :: Int ..] balances
    mapM_ (\(i, b) -> printf "    Account %d: %6d\n" i b) labeled

-----------------------------------------------------------------------------
-- Main
-----------------------------------------------------------------------------

main :: IO ()
main = do
    putStrLn "=============================================="
    putStrLn "  STM Bank Transfer Demo"
    putStrLn "  Composable concurrent transactions"
    putStrLn "=============================================="
    putStrLn ""

    -- Create 5 accounts with varying balances
    a1 <- newAccount 1000
    a2 <- newAccount 2000
    a3 <- newAccount 1500
    a4 <- newAccount 3000
    a5 <- newAccount 500

    let accounts = [a1, a2, a3, a4, a5]

    putStrLn "Initial balances:"
    printBalances accounts
    initTotal <- totalBalance accounts
    putStrLn $ "  -------------------"
    printf "  Total: %6d\n" initTotal
    putStrLn ""

    -- Spawn 7 concurrent transfer threads
    let numThreads = 7
        transfersPerThread = 5

    putStrLn $ "Spawning " ++ show numThreads ++ " concurrent transfer threads"
    putStrLn $ "(each performing " ++ show transfersPerThread ++ " random transfers)..."
    putStrLn ""

    -- Use MVars to synchronize completion
    doneMVars <- replicateM numThreads $ do
        done <- newEmptyMVar
        _ <- forkIO $ transferThread accounts transfersPerThread >> putMVar done ()
        return done

    -- Wait for all threads to finish
    putStrLn "Waiting for all transfers to complete..."
    mapM_ takeMVar doneMVars
    putStrLn ""

    -- Print final state
    putStrLn "Final balances:"
    printBalances accounts
    finalTotal <- totalBalance accounts
    putStrLn $ "  -------------------"
    printf "  Total: %6d\n" finalTotal
    putStrLn ""

    -- Verify conservation
    putStrLn "Verification:"
    if initTotal == finalTotal
        then do
            putStrLn "  Total balance is CONSERVED."
            putStrLn "  STM guarantees atomicity and consistency."
            putStrLn ""
            putStrLn "  Key insight: Because each transfer is a single STM"
            putStrLn "  transaction, the system is always in a consistent"
            putStrLn "  state. No lock ordering, no deadlocks, no races."
        else
            putStrLn "  ERROR: Total balance changed -- inconsistency detected!"

File: lua/stm/stm_bank.lua
--[[
  stm_bank.lua
  Simulate STM (Software Transactional Memory) using coroutines.

  Implements TVar-like transactional variables with versioning,
  atomically() for transactions, and retry on conflict.
  Compare with: Python stm_model.py, Haskell STMBank.hs
]]

local STM = {}
STM.__index = STM

-- TVar: transactional variable with versioning
local TVar = {}
TVar.__index = TVar

function TVar.new(value)
  return setmetatable({
    value = value,
    version = 0
  }, TVar)
end

-- Transaction: tracks reads and writes
local Transaction = {}
Transaction.__index = Transaction

function Transaction.new()
  return setmetatable({
    read_set = {},  -- tvar -> {value, version}
    write_set = {}, -- tvar -> new_value
    should_retry = false
  }, Transaction)
end

function Transaction:read_tvar(tvar)
  if self.write_set[tvar] ~= nil then
    return self.write_set[tvar]
  end
  self.read_set[tvar] = { value = tvar.value, version = tvar.version }
  return tvar.value
end

function Transaction:write_tvar(tvar, value)
  self.write_set[tvar] = value
end

function Transaction:validate()
  for tvar, snapshot in pairs(self.read_set) do
    if tvar.version ~= snapshot.version then
      return false
    end
  end
  return true
end

function Transaction:commit()
  for tvar, new_value in pairs(self.write_set) do
    tvar.value = new_value
    tvar.version = tvar.version + 1
  end
end

-- Global transaction execution
function atomically(fn)
  while true do
    local txn = Transaction.new()
    local ok, result = pcall(fn, txn)

    if not ok then
      -- Retry on conflict
      coroutine.yield("retry")
    else
      if txn:validate() then
        txn:commit()
        return result
      end
    end
    -- Yield before retry
    coroutine.yield("conflict")
  end
end

-- Application: bank transfer with STM
local function bank_transfer(from, to, amount)
  return atomically(function(txn)
    local from_bal = txn:read_tvar(from)
    local to_bal = txn:read_tvar(to)

    if from_bal < amount then
      error("insufficient funds")
    end

    txn:write_tvar(from, from_bal - amount)
    txn:write_tvar(to, to_bal + amount)
    return string.format("Transferred %d", amount)
  end)
end

-- Demo
print("=== STM: Bank Transfer Demo (Lua) ===\n")

local acc_a = TVar.new(1000)
local acc_b = TVar.new(500)
local total_initial = acc_a.value + acc_b.value

print(string.format("Initial: A=%d, B=%d, total=%d\n", acc_a.value, acc_b.value, total_initial))

-- Create concurrent transaction coroutines
local transactions = {}
for i = 1, 5 do
  table.insert(transactions, coroutine.create(function()
    for j = 1, 3 do
      local result = bank_transfer(acc_a, acc_b, 50)
      print("  " .. result)
      coroutine.yield()
    end
  end))
end

-- Round-robin scheduler
local running = true
while running do
  running = false
  for _, co in ipairs(transactions) do
    local status = coroutine.status(co)
    if status == "suspended" or status == "running" then
      running = true
      local ok, err = coroutine.resume(co)
      if not ok and err ~= "retry" and err ~= "conflict" then
        print("  Transaction error: " .. tostring(err))
      end
    end
  end
end

local total_final = acc_a.value + acc_b.value
print(string.format("\nFinal: A=%d, B=%d, total=%d", acc_a.value, acc_b.value, total_final))
print(string.format("Total conserved: %s", (total_final == total_initial) and "YES" or "NO"))

print("\nSTM demo passed.")
